File: docs/js/Core1/Inventory-Management/DailyDemand.js
class DailyDemand extends eVENT {
  constructor({occTime, delay, quantity, shop}) {
    super({occTime, delay});
    this.quantity = quantity;
    this.shop = shop;
  }
  onEvent() {
    var q = this.quantity,
        prevStockLevel = this.shop.stockQuantity;
    // update lostSales if demand quantity greater than stock level
    if (q > prevStockLevel) {
      // increment the stock-out counter by 1
      sim.stat.nmrOfStockOuts++;
      // increment the "lostSales" statistics variable by the missing quantity
      sim.stat.lostSales += q - prevStockLevel;
    }
    // update stockQuantity
    this.shop.stockQuantity = Math.max( prevStockLevel-q, 0);
    switch (sim.model.p.reviewPolicy) {
    case "continuous":
      // schedule Delivery if stock level falls below reorder level
      if (prevStockLevel > this.shop.reorderPoint &&
          prevStockLevel - q <= this.shop.reorderPoint) {
        return [new Delivery({
          delay: Delivery.leadTime(),
          quantity: this.shop.targetInventory - this.shop.stockQuantity,
          receiver: this.shop
        })];
      } else return [];  // no follow-up events
    case "periodic":
      // periodically schedule new Delivery events
      if (sim.time % this.shop.reorderInterval === 0) {
        return [new Delivery({
          delay: Delivery.leadTime(),
          quantity: this.shop.targetInventory - this.shop.stockQuantity,
          receiver: this.shop
        })];
      } else return [];  // no follow-up events
    }
  }
  static quantity() {
    return rand.uniformInt( 5, 30);
  }
  static recurrence() {
    return 1;  // each day
  }
  createNextEvent() {
    return new DailyDemand({
      delay: DailyDemand.recurrence(),
      quantity: DailyDemand.quantity(),
      shop: this.shop
    });
  }
}
DailyDemand.labels = {"quantity":"quant"};

File: Java/Core1/src/main/java/de/oes/core1/inventorymanagment/DailyDemand.java
package de.oes.core1.inventorymanagment;

import java.util.ArrayList;
import java.util.List;

import de.oes.core1.foundations.ExogenousEvent;
import de.oes.core1.foundations.eVENT;
import de.oes.core1.lib.MathLib;
import de.oes.core1.sim.Simulator;


public class DailyDemand extends ExogenousEvent {
	private Integer quantity;
	private SingleProductShop shop;
	
	public DailyDemand(Simulator sim, Number occTime, Number delay, Integer quantity, SingleProductShop shop) {
		super(sim, occTime, delay);
		this.shop = shop;
		this.quantity = quantity;
	}

	@Override
	public List<eVENT> onEvent() {
		Integer q = this.quantity;
		Integer prevStockLevel = this.shop.getQuantityInStock();
		if(q > prevStockLevel) {
			this.getSim().incrementStat("nmrOfStockOuts", 1);
			this.getSim().incrementStat("lostSales", q - prevStockLevel);
		}
		this.shop.setQuantityInStock(Math.max(prevStockLevel - q, 0));
		if((this.getSim().getModel().getP().get("reviewPolicy")).equals("continuous")) {
			if(prevStockLevel > this.shop.getReorderPoint() &&
					prevStockLevel -q <= this.shop.getReorderPoint()) {
				return List.of(
						new Delivery(
								this.getSim(), 
								null, // no occ. time
								Delivery.leadTime().longValue(), 
								this.shop.getTargetInventory() - this.shop.getQuantityInStock(), 
								this.shop)
						); 
			} else return new ArrayList<eVENT>();
		} else { // periodic
			 // periodically schedule new Delivery events
			if((this.getSim().getTime() % this.shop.getReorderInterval()) == 0) {
				return List.of(
						new Delivery(
								this.getSim(), 
								null, // no occ. time
								Delivery.leadTime().longValue(), 
								this.shop.getTargetInventory() - this.shop.getQuantityInStock(), 
								this.shop)
						); 
			} else return new ArrayList<eVENT>();
		}
	}

	
	public static Integer quantity() {
		return MathLib.getUniformRandomInteger(5, 30);
	}
	
	public Long reccurence() {
		return 1l;
	}
	
	public DailyDemand createNextEvent() {
		return new DailyDemand(
				this.getSim(), 
				null, // occ. time
				this.reccurence(), // delay
				DailyDemand.quantity(), 
				this.shop);
	}
	
	@Override
	public String toString() {
		return this.getClass().getSimpleName() + "{quant : " + this.quantity + "} @" + this.getOccTime();
	}
}
